lib/ansible/modules/cloud/podman/podman_container.py

```diff
@@ -0,0 +1,1281 @@
+        return ['stop', params['name']]
+    if action == 'delete':
+        return ['rm', '-f', params['name']]
+    if action in ['create', 'run']:
+        cmd = ['run', '--name', params['name']]
+
+        if params['detach']:
+            cmd += ['--detach']
+
+        for host_ip in params['etc_hosts'].items():
+            cmd += ['--add-host', ':'.join(host_ip)]
+
+        for annotate in params['annotation'].items():
+            cmd += ['--annotation=' + '='.join(annotate)]
+        # Another thousand parameters
+
+        if params['blkio_weight'] is not None:
+            cmd += ['--blkio-weight', params['blkio_weight']]
+
+        if params['blkio_weight_device']:
+            cmd += ['--blkio-weight-device', params['blkio_weight_device']]
+
+        for cap_add in params['cap_add']:
+            cmd += ['--cap-add', cap_add]
+
+        for cap_drop in params['cap_drop']:
+            cmd += ['--cap-drop', cap_drop]
+
+        if params['cgroup_parent']:
+            cmd += ['--cgroup-parent', params['cgroup_parent']]
+
+        if params['cidfile']:
+            cmd += ['--cidfile', params['cidfile']]
+
+        if params['conmon_pidfile']:
+            cmd += ['--conmon-pidfile', params['conmon_pidfile']]
+
+        if params['cpu_period']:
+            cmd += ['--cpu-period', params['cpu_period']]
+
+        if params['cpu_rt_runtime'] is not None:
+            cmd += ['--cpu-rt-runtime', params['cpu_rt_runtime']]
+
+        if params['cpu_shares'] is not None:
+            cmd += ['--cpu-shares', params['cpu_shares']]
+
+        if params['cpus']:
+            cmd += ['--cpus', params['cpus']]
+
+        if params['cpuset_cpusr']:
+            cmd += ['--cpuset-cpusr', params['cpuset_cpusr']]
+
+        if params['cpuset_mems']:
+            cmd += ['--cpuset-mems', params['cpuset_mems']]
+
+        if params['detach_keys']:
+            cmd += ['--detach-keys', params['detach_keys']]
+
+        if params['device']:
+            cmd += ['--device', params['device']]
+
+        if params['device_read_bps']:
+            cmd += ['--device-read-bps', params['device_read_bps']]
+
+        if params['device_read_iops']:
+            cmd += ['--device-read-iops', params['device_read_iops']]
+
+        if params['device_write_bps']:
+            cmd += ['--device-write-bps', params['device_write_bps']]
+
+        if params['device_write_iops']:
+            cmd += ['--device-write-iops', params['device_write_iops']]
+
+        if params['dns']:
+            cmd += ['--dns=' + ','.join(params['dns'])]
+
+        if params['dns_option']:
+            cmd += ['--dns-option', params['dns_option']]
+
+        if params['dns_search'] is not None:
+            cmd += ['--dns-search=' + params['dns_search']]
+
+        if params['entrypoint']:
+            cmd += ['--entrypoint', params['entrypoint']]
+
+        for env_value in params['env'].items():
+            cmd += ['--env', "=".join([str(k) for k in env_value])]
```

Comment:
All of these uses of `str` should be switched to `to_bytes(var, errors='surrogate_or_strict')`.

`to_bytes` comes from `ansible.module_utils._text`.

Reply:
Nope, not `to_bytes` but `to_native` in this case. Or maybe `to_text`.

@abadger right?

Reply:
I explicitly reached out to @abadger for confirmation earlier:

```
sivel [4:15 PM]
@abadger1999 when shelling out via Popen, should we be using bytes?  I'm being lazy and just asking, instead of investigating :slightly_smiling_face:

abadger1999 [4:15 PM]
Yes
subprocess.Popen() uses bytes.

sivel [4:16 PM]
twas what I thought

abadger1999 [4:16 PM]
bytes in bytes out.
python will try to translate text in into bytes but on Python2 especially, it can get it wrong.
```

Reply:
Thanks for the confirmation!

Reply:
In the end I run all over this list with ``str`` anyway, so I'll just will change there to use ``to_bytes(var, errors='surrogate_or_strict')`` instead of ``str``. Is it ok? In the end of ``construct_command_from_params`` function.

Reply:
Yes, but it's also worth prefixing literals with `b`

Reply:
Oh, and in places with variables where you concatenate that with literals, also apply this function to avoid certain edge-cases which appear sometimes when you try to merge different data types.

Reply:
I'm afraid that writing ``to_bytes(var, errors='surrogate_or_strict')`` near every string will make the code completely unreadable.. I'd like to convert it as late as possible

Reply:
Yes, you can convert resulting strings. But when concatenating things it may result in unpleasant side-effects.
```python
# WRONG
b'some str %s' % other_var  # <-- Python (esp Python 2) tries to turn `other_var` into bytes, sometimes incorrectly which may even cause exceptions in obscure cases
```
```python
# CORRECT
b'some str %s' % to_bytes(other_var, errors='surrogate_or_strict')  # when you convert it explicitly, it'll always succeed
```

If you don't like this `errors='surrogate_or_strict'` part, you can use curring:
```python
from functools import partial
...
from ansible.module_utils._text import to_bytes
...
to_bytes_sos = partial(to_bytes, errors='surrogate_or_strict')
#   ^
#   |
#   |
# somewhere at the top of the module
...
# And everywhere else in code:
#   |
#   |
#   v
b'some str %s' % to_bytes_sos(other_var)
```

Hope this helps.

Reply:
@webknjaz correct, that's why I don't set "b" near strings that I manipulate with them.
If it will be just as it's now:
```
'some str %s' % other_var 
```
it should work for both python 2 and 3, right? And then in the end I convert aaallll this to bytes. Is it correct or I miss something?

Reply:
You don't always have a guarantee of `other_var` being the same type. And `'some str %s'` has different types under Python 2 and Python 3. That's why it's important to explicitly use the correct literal.

Reply:
Well, in this code I always have
```'some str %s' % str(other_var)``` 
so it will guarantee we deal with the same type, also it's ensured we have``str`` type by defining it in the params ``argument_spec``. So it's **2** levels of guards here.
That's right that it has different types in Py2 and Py3, but in the same moment it runs with one of them only, it can't run half of code with different Py version. And this code is good for both. I tested it on systems with both default Py2 and Py3.
That's not that I'm lazy to add a function, but this part of code will be most changing in this module I suppose. The podman project is quite raw and fast changing, so most of changes I'd expect to be with parameters, defaults, their manipulations etc - what is defined in this piece of code. I don't expect something to be changed in containers start logic..
For being this maintainable and fixable by other people, not always familiar with Ansible and its internal kitchen I'd like to keep it as simple as possible. Adding a simple variable, or changing one will be much harder with all these non obvious wrappers and may lead to all kinds of issues.

Reply:
Please don't use `str()` because it has the same problems as not using it. `to_bytes` is specifically designed to properly deal with corner cases. And we don't just use `str()` exactly because in certain corner cases which are hard to reproduce if you don't know what they are.

Reply:
Python 3 example of why it may be harmful:
```python
>>> str(b'test')
"b'test'"
```

Reply:
This discussion is taking a bunch of twists and turns but I believe @sshnaidm has the right idea except with regard to using ```str()```.

In module code the general assumption is that all strings are native strings (unadorned literal strings like ```"native string"```) rather than text strings (```u"text string"```) or byte strings (```b"byte string"```).  Unless otherwise noted, AnsibleModule methods should return native strings, for instance.  This is not always the case (because other APIs might return text strings or byte strings specifically.  For instance, subprocess.Popen() always deals in byte strings.) but we decided that standardizing on anything besides native strings would cause issues for all of the modules which were written before a standard was decided upon.

So wrt ```'some str %s' % str(other_var)```, it's extraneous to run ```str()``` on the parameters parsed by the module.  Since you're combining with a native format string, you should just combine the two without using ```str()``` => ```'some str %s' % other_var```.  If you are ever concerned that other_var contains something that is not a string and could cause a traceback when converted by the format string, the way to convert it (in module code) would be ```some str %s' % to_native(other_var, errors='surrogate_or_strict')```.  The reason is that str() on Python2 will traceback if it ends up with non-ascii characters and str() on Python3 can end up repr'ing  unexpectedly:
```
[pts/7@peru ~]$ python2 -c 'print(str(u"café"))'                                                          (07:14:52)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode characters in position 3-4: ordinal not in range(128)
[pts/6@peru ~]$ python3 -c 'print(str(u"café".encode("utf-8")))'                                          (07:15:34)
b'caf\xc3\xa9'
```

Whether you need to use ```to_bytes()``` at all... I should have asked for more information last night.... ```module.run_command()``` is a wrapper around ```subprocess.Popen()``` and one of the things it's supposed to do is take native strings in and return native strings out rather than doing bytes-in and bytes-out as Popen does.  So if everything is working right, you shouldn't have to convert to byte strings using ```to_bytes()```.  however, I did just look at the code to ```module.run_command()``` and I see that there are a few code paths where native strings wouldn't be converted to byte strings when using module.run_command() so byte strings are safer, but only because of a bug in ```run_command()```.

One last thing I saw in this discussion.
```
b'some str %s' % other_var  # <-- Python (esp Python 2) tries to turn `other_var` into bytes, sometimes incorrectly which may even cause exceptions in obscure cases
```

That's actually unsafe on Python3, not on Python2.  On Python2, percent-format strings will always coerce to a text string if at least one of the pieces is text.  So if ```other_var = u'café'``` output of that will be u'some str café'.  On Python3, Python will error if you try to use a percent-formatted byte string with a text variable as it won't automatically coerce the values.
```
[pts/6@peru ~]$ python3 -c 'b"some str %s" % u"café"'                                                     (07:23:17)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: %b requires a bytes-like object, or an object that implements __bytes__, not 'str'
```
Note that ```"string".formst()``` behaviour is different.  If you want information on that, I'd suggest reading:https://anonbadger.wordpress.com/2016/01/05/python2-string-format-and-unicode/

Reply:
Thanks for explanations, folks. I removed all str() from this file, with string formatting as well. Only one that left change to to_bytes function. Also added some non-string values to tests to be sure we're good with that.